Added frame duration per state & fixed flipping

0xnicholasc opened the pull request:
First time contributing to another project. Hope you find this useful.

0xnicholasc committed:
Added ability to set different frame durations per state

0xnicholasc committed:
Fixed Horizontal & Veritcal Flip 

Old flipping math made the sprite change position on the screen while flipping

0xnicholasc committed:
Removed scaling from flip math

Translation while flipping does not require to be multiplied by scale. This is handled by t.Scale. This creates expected flipping behavior

0xnicholasc commented:
Previous flipping logic only worked when Scale = 1. If scale was >1, Sprite is moved when flipped.

matt-hayward commented on olcPGEX_AnimatedSprite.h:
```diff
@@ -69,16 +69,17 @@ namespace olc
 		void Draw(float fElapsedTime, olc::vf2d position);
-		void AddState(std::string stateName, std::vector<std::string> imagePaths);
-		void AddState(std::string stateName, std::vector<olc::vi2d> spriteLocations);
+		void AddState(std::string stateName, float frameDuration, std::vector<std::string> imagePaths);
```
I think it would make sense to have these as additional overloads to the `AddState` method - that way the default can still be used without having to define it for every single state if it isn't needed.

0xnicholasc replied:
I've added that in the last commits.

matt-hayward commented on olcPGEX_AnimatedSprite.h:
```diff
@@ -191,7 +198,8 @@ namespace olc
 		if (scale <= 0.0f) {
 			spriteScale = 1.0f;
-		} else {
+		}
```
Revert formatting here, please.

matt-hayward commented on olcPGEX_AnimatedSprite.h:
```diff
@@ -201,10 +209,11 @@ namespace olc
-		} else if (flip == FLIP_MODE::VERTICAL) {
-			t.Translate(0, -((spriteSize.y / 2) * spriteScale));
+		}
+		else if (flip == FLIP_MODE::VERTICAL) {
```
Revert formatting here, please.

matt-hayward commented on olcPGEX_AnimatedSprite.h:
```diff
@@ -98,8 +99,9 @@ namespace olc
+		std::map<std::string, float> frameDurations;
 		float frameTimer = 0.0f;
-		float frameDuration = 0.1f;
+		//float frameDuration = 0.1f;
```
I think this should be kept as a default value (see above comment).

0xnicholasc replied:
Added it back, this time as a constant.

matt-hayward commented on olcPGEX_AnimatedSprite.h:
```diff
@@ -115,7 +117,7 @@ namespace olc
 		frameTimer += fElapsedTime;
 
-		if (frameTimer >= frameDuration) {
+		if (frameTimer >= frameDurations[state]) {
```
I think this should handle states that use the default duration.

0xnicholasc replied:
The default duration will be set into the frameDurations array when the state is added.

matt-hayward commented on olcPGEX_AnimatedSprite.h:
```diff
@@ -131,7 +133,7 @@ namespace olc
 		frameTimer += fElapsedTime;
 
-		if (frameTimer >= frameDuration) {
+		if (frameTimer >= frameDurations[state]) {
```
I think this should handle states that use the default duration.

matt-hayward commented on olcPGEX_AnimatedSprite.h:
```diff
@@ -69,16 +69,17 @@ namespace olc
 		void SetSpriteSize(olc::vi2d size);
 		olc::vi2d GetSpriteSize();
 		void SetSpriteScale(float scale);
+		void SetFrameDuration(float duration);
```
This isn't implemented anywhere?

0xnicholasc replied:
This is a vestige of something else I was playing with. Removed.

matt-hayward commented:
Hey, thanks for contributing! Yeah, this looks useful and was something I was thinking about.

I've added some feedback for you, so let me know if you have any questions.

For the flip issue, I tested on 2x scale and it seemed to work fine. Perhaps the actual image you're using isn't quite centered on the canvas?

If possible, could you also create an issue for the two things fixed here?

0xnicholasc commented:
Thanks for the feedback, I'll fix those things you mentioned tomorrow!

On Wed, Jan 29, 2020, 11:13 PM matt-hayward <notifications@github.com>
wrote:

> Hey, thanks for contributing! Yeah, this looks useful and was something I
> was thinking about.
>
> I've added some feedback for you, so let me know if you have any questions.
>
> For the flip issue, I tested on 2x scale and it seemed to work fine.
> Perhaps the actual image you're using isn't quite centered on the canvas?
>
> If possible, could you also create an issue for the two things fixed here?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/matt-hayward/olcPGEX_AnimatedSprite/pull/2?email_source=notifications&email_token=AONHSGXKS233HBGHD2BQ5ZDRAJHXPA5CNFSM4KNOR6NKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEKJT4DI#issuecomment-580075021>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AONHSGT3XRSDS4HF6KKJBRTRAJHXPANCNFSM4KNOR6NA>
> .
>

0xnicholasc committed:
Default frame duration

I restored the default frame duration and added in overrides to accommodate it. Fixed formatting for else if, removed unimplemented prototype.

matt-hayward commented on olcPGEX_AnimatedSprite.h:
```diff
@@ -98,8 +100,9 @@ namespace olc
 		std::map<std::string, std::vector<olc::Sprite*>> multiFrames;
 		std::map<std::string, std::vector<olc::vi2d>> singleFrames;
+		std::map<std::string, float> frameDurations;
+		const float DEFAULT_FRAME_DURATION = 0.1f;
```
I don't really like the idea of this being a const. I'm happy for it to stay public, though, so that the default 'global' frame duration can be set. As mentioned, I was thinking about letting that be changed anyway.

matt-hayward commented:
I've added one final bit of feedback, then I think everything is fine with this PR. Could you create an issue for this, though, please?

0xnicholasc commented:
> I've added one final bit of feedback, then I think everything is fine with this PR. Could you create an issue for this, though, please?

Incorporating your feedback right now. Did you want me to create an issue on the olcPixelGameEngine?

0xnicholasc committed:
Changes to default frame duration

Made the default change duration mutable & public.

matt-hayward commented:
> > I've added one final bit of feedback, then I think everything is fine with this PR. Could you create an issue for this, though, please?
> 
> Incorporating your feedback right now. Did you want me to create an issue on the olcPixelGameEngine?

No, create an issue on this repo, please.

0xnicholasc commented:
I did yesterday.

On Sat, Feb 1, 2020, 9:22 AM matt-hayward <notifications@github.com> wrote:

> I've added one final bit of feedback, then I think everything is fine with
> this PR. Could you create an issue for this, though, please?
>
> Incorporating your feedback right now. Did you want me to create an issue
> on the olcPixelGameEngine?
>
> No, create an issue on this repo, please.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/matt-hayward/olcPGEX_AnimatedSprite/pull/2?email_source=notifications&email_token=AONHSGSVHD5LZZEZFIE2GS3RAWASXA5CNFSM4KNOR6NKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEKQ6GOI#issuecomment-581034809>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AONHSGWQNFFGSPBGXSCCZ4LRAWASXANCNFSM4KNOR6NA>
> .
>

matt-hayward commented:
I've created two separate issues for this PR:

https://github.com/matt-hayward/olcPGEX_AnimatedSprite/issues/4

https://github.com/matt-hayward/olcPGEX_AnimatedSprite/issues/5
